File: experiment-14/runs/language=java_model=opus-4.8-fast_prompt=ATDD/rep1/src/main/java/com/brazilsoccer/mcp/DataStore.java
package com.brazilsoccer.mcp;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVParser;
import org.apache.commons.csv.CSVRecord;
import org.apache.commons.csv.DuplicateHeaderMode;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Loads all provided CSV datasets into in-memory collections.
 *
 * <p>Several files overlap (the Brasileirão appears in three of them). Matches are
 * de-duplicated by {@link Match#dedupeKey()}, keeping the first source loaded. To
 * avoid double-counting league tables, the extended statistics file does not
 * contribute matches for any (competition, season) already covered by a primary
 * source — it only extends coverage (e.g. Série B/C, newer seasons).
 */
public final class DataStore {

    private static final Pattern HAS_STATE_SUFFIX = Pattern.compile("[\\s-]+[A-Z]{2,3}\\s*$");

    private final List<Match> matches;
    private final List<Player> players;

    private DataStore(List<Match> matches, List<Player> players) {
        this.matches = matches;
        this.players = players;
    }

    public List<Match> matches() {
        return matches;
    }

    public List<Player> players() {
        return players;
    }

    public static DataStore load(Path dataDir) {
        List<Match> matches = new ArrayList<>();
        Set<String> seen = new HashSet<>();

        // Sources are loaded in priority order. Each later source is gated against the
        // (competition, season) coverage already established, so overlapping datasets
        // (the Brasileirão appears in three files) never double-count a league.
        loadBrasileirao(dataDir.resolve("Brasileirao_Matches.csv"), matches, seen, Set.of());
        loadNovo(dataDir.resolve("novo_campeonato_brasileiro.csv"), matches, seen, coverageOf(matches));
        loadLibertadores(dataDir.resolve("Libertadores_Matches.csv"), matches, seen, coverageOf(matches));
        loadCup(dataDir.resolve("Brazilian_Cup_Matches.csv"), matches, seen, coverageOf(matches));
        loadExtended(dataDir.resolve("BR-Football-Dataset.csv"), matches, seen, coverageOf(matches));

        List<Player> players = loadPlayers(dataDir.resolve("fifa_data.csv"));
        return new DataStore(matches, players);
    }

    /** Snapshot of the (competition, season) pairs already loaded. */
    private static Set<String> coverageOf(List<Match> matches) {
        Set<String> coverage = new LinkedHashSet<>();
        for (Match m : matches) {
            if (m.season > 0) {
                coverage.add(m.competition + "|" + m.season);
            }
        }
        return coverage;
    }

    // ---- Match loaders -------------------------------------------------------

    private static void loadBrasileirao(Path file, List<Match> out, Set<String> seen, Set<String> covered) {
        forEachRecord(file, record -> {
            int season = parseSeason(get(record, "season"), get(record, "datetime"));
            Match m = buildMatch(Competition.SERIE_A, season,
                    parseDate(get(record, "datetime")), get(record, "round"),
                    get(record, "home_team"), get(record, "home_team_state"),
                    get(record, "away_team"), get(record, "away_team_state"),
                    get(record, "home_goal"), get(record, "away_goal"));
            add(m, out, seen, covered);
        });
    }

    private static void loadNovo(Path file, List<Match> out, Set<String> seen, Set<String> covered) {
        forEachRecord(file, record -> {
            int season = parseSeason(get(record, "Ano"), get(record, "Data"));
            Match m = buildMatch(Competition.SERIE_A, season,
                    parseDate(get(record, "Data")), get(record, "Rodada"),
                    get(record, "Equipe_mandante"), get(record, "Mandante_UF"),
                    get(record, "Equipe_visitante"), get(record, "Visitante_UF"),
                    get(record, "Gols_mandante"), get(record, "Gols_visitante"));
            add(m, out, seen, covered);
        });
    }

    private static void loadLibertadores(Path file, List<Match> out, Set<String> seen, Set<String> covered) {
        forEachRecord(file, record -> {
            int season = parseSeason(get(record, "season"), get(record, "datetime"));
            Match m = buildMatch(Competition.LIBERTADORES, season,
                    parseDate(get(record, "datetime")), get(record, "stage"),
                    get(record, "home_team"), null,
                    get(record, "away_team"), null,
                    get(record, "home_goal"), get(record, "away_goal"));
            add(m, out, seen, covered);
        });
    }

    private static void loadCup(Path file, List<Match> out, Set<String> seen, Set<String> covered) {
        forEachRecord(file, record -> {
            int season = parseSeason(get(record, "season"), get(record, "datetime"));
            Match m = buildMatch(Competition.COPA_DO_BRASIL, season,
                    parseDate(get(record, "datetime")), get(record, "round"),
                    get(record, "home_team"), null,
                    get(record, "away_team"), null,
                    get(record, "home_goal"), get(record, "away_goal"));
            add(m, out, seen, covered);
        });
    }

    private static void loadExtended(Path file, List<Match> out, Set<String> seen, Set<String> covered) {
        forEachRecord(file, record -> {
            Competition comp = Competition.fromTournamentLabel(get(record, "tournament"));
            if (comp == null) {
                return;
            }
            String date = parseDate(get(record, "date"));
            int season = parseSeason(null, date);
            Match m = buildMatch(comp, season, date, null,
                    get(record, "home"), null,
                    get(record, "away"), null,
                    get(record, "home_goal"), get(record, "away_goal"));
            add(m, out, seen, covered);
        });
    }

    private static void add(Match m, List<Match> out, Set<String> seen, Set<String> covered) {
        if (m == null) {
            return;
        }
        // Skip a (competition, season) already owned by a higher-priority source.
        if (m.season > 0 && covered.contains(m.competition + "|" + m.season)) {
            return;
        }
        if (!seen.add(m.dedupeKey())) {
            return;
        }
        out.add(m);
    }

    private static Match buildMatch(Competition comp, int season, String date, String round,
                                    String homeRaw, String homeUf, String awayRaw, String awayUf,
                                    String homeGoalRaw, String awayGoalRaw) {
        int hg = parseGoals(homeGoalRaw);
        int ag = parseGoals(awayGoalRaw);
        if (hg < 0 || ag < 0) {
            return null; // unplayed / malformed row
        }
        String round0 = blankToNull(round);
        return new Match(comp, season, date, round0,
                display(homeRaw, homeUf), display(awayRaw, awayUf),
                TeamNames.matchKey(homeRaw), TeamNames.matchKey(awayRaw),
                TeamNames.identityKey(homeRaw, homeUf), TeamNames.identityKey(awayRaw, awayUf),
                hg, ag);
    }

    /** Human-readable team label, always including the state code when known. */
    static String display(String rawName, String stateColumn) {
        String cleaned = rawName == null ? "" : rawName.replaceAll("\\s*\\([^)]*\\)\\s*", " ").trim();
        if (HAS_STATE_SUFFIX.matcher(cleaned).find()) {
            return cleaned;
        }
        if (stateColumn != null) {
            String uf = stateColumn.trim();
            if (!uf.isEmpty() && !uf.equalsIgnoreCase("na")) {
                return cleaned + "-" + uf.toUpperCase(Locale.ROOT);
            }
        }
        return cleaned;
    }

    // ---- Player loader -------------------------------------------------------

    private static List<Player> loadPlayers(Path file) {
        List<Player> players = new ArrayList<>();
        forEachRecord(file, record -> {
            String name = get(record, "Name");
            if (name == null || name.isBlank()) {
                return;
            }
            players.add(new Player(
                    name,
                    parseInt(get(record, "Age"), 0),
                    orEmpty(get(record, "Nationality")),
                    parseInt(get(record, "Overall"), 0),
                    parseInt(get(record, "Potential"), 0),
                    orEmpty(get(record, "Club")),
                    orEmpty(get(record, "Position")),
                    orEmpty(get(record, "Jersey Number"))));
        });
        return players;
    }

    // ---- CSV plumbing --------------------------------------------------------

    private interface RecordConsumer {
        void accept(CSVRecord record);
    }

    private static void forEachRecord(Path file, RecordConsumer consumer) {
        CSVFormat format = CSVFormat.DEFAULT.builder()
                .setHeader()
                .setSkipHeaderRecord(true)
                .setIgnoreSurroundingSpaces(true)
                .setAllowMissingColumnNames(true)
                .setDuplicateHeaderMode(DuplicateHeaderMode.ALLOW_ALL)
                .build();
        try (BufferedReader reader = newBomSafeReader(file);
             CSVParser parser = CSVParser.parse(reader, format)) {
            for (CSVRecord record : parser) {
                consumer.accept(record);
            }
        } catch (IOException e) {
            throw new UncheckedIOException("Failed to read " + file, e);
        }
    }

    private static BufferedReader newBomSafeReader(Path file) throws IOException {
        BufferedReader reader = Files.newBufferedReader(file, StandardCharsets.UTF_8);
        reader.mark(1);
        if (reader.read() != 0xFEFF) {
            reader.reset();
        }
        return reader;
    }

    private static String get(CSVRecord record, String column) {
        if (!record.isMapped(column)) {
            return null;
        }
        String v = record.get(column);
        return v == null ? null : v.trim();
    }

    // ---- Parsing helpers -----------------------------------------------------

    static int parseGoals(String raw) {
        if (raw == null || raw.isBlank()) {
            return -1;
        }
        try {
            return (int) Math.round(Double.parseDouble(raw.trim()));
        } catch (NumberFormatException e) {
            return -1;
        }
    }

    static int parseInt(String raw, int fallback) {
        if (raw == null || raw.isBlank()) {
            return fallback;
        }
        try {
            return (int) Math.round(Double.parseDouble(raw.trim()));
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    static int parseSeason(String seasonRaw, String dateRaw) {
        int s = parseInt(seasonRaw, -1);
        if (s > 0) {
            return s;
        }
        String date = parseDate(dateRaw);
        if (date != null && date.length() >= 4) {
            return parseInt(date.substring(0, 4), -1);
        }
        return -1;
    }

    /** Normalizes the various date formats to ISO yyyy-MM-dd. */
    static String parseDate(String raw) {
        if (raw == null) {
            return null;
        }
        String v = raw.trim();
        if (v.isEmpty()) {
            return null;
        }
        if (v.contains("/")) {
            String[] parts = v.split("/");
            if (parts.length == 3 && parts[2].length() == 4) {
                return String.format("%s-%02d-%02d",
                        parts[2], parseInt(parts[1], 0), parseInt(parts[0], 0));
            }
            return null;
        }
        String datePart = v.contains(" ") ? v.substring(0, v.indexOf(' ')) : v;
        return datePart.length() >= 10 ? datePart.substring(0, 10) : datePart;
    }

    private static String blankToNull(String s) {
        return (s == null || s.isBlank()) ? null : s;
    }

    private static String orEmpty(String s) {
        return s == null ? "" : s;
    }
}

File: experiments/adrianco/experiment-15-sonnet5/brazil/runs/language=typescript_model=opus-4.8_prompt=bdd/rep1/src/dataStore.ts
/**
 * The DataStore loads every provided CSV file once and exposes the normalized
 * in-memory collections (matches + players) that all queries read from.
 *
 * Loading all six files into memory keeps every query well under the 2s/5s
 * performance targets in the spec: the datasets total ~40k matches and ~18k
 * players, which is trivial to scan in memory.
 */
import { existsSync } from "node:fs";
import { join } from "node:path";
import { loadCsv, type CsvRow } from "./csv.js";
import { cleanTeamName, parseDate, parseGoals, parseIntOrNull } from "./normalize.js";
import type { Match, Player } from "./types.js";

/** Canonical Brasileirão league name used for league-table calculations. */
export const BRASILEIRAO = "Brasileirão Série A";
export const COPA_DO_BRASIL = "Copa do Brasil";
export const LIBERTADORES = "Copa Libertadores";

const MATCH_FILES = {
  brasileirao: "Brasileirao_Matches.csv",
  cup: "Brazilian_Cup_Matches.csv",
  libertadores: "Libertadores_Matches.csv",
  extended: "BR-Football-Dataset.csv",
  historical: "novo_campeonato_brasileiro.csv",
} as const;

const PLAYER_FILE = "fifa_data.csv";

export class DataStore {
  readonly matches: Match[] = [];
  readonly players: Player[] = [];
  private loaded = false;

  constructor(private readonly dataDir: string) {}

  /** Load and normalize all datasets. Idempotent. */
  load(): void {
    if (this.loaded) return;
    this.loadBrasileirao();
    this.loadCup();
    this.loadLibertadores();
    this.loadExtended();
    this.loadHistorical();
    this.loadPlayers();
    this.loaded = true;
  }

  private filePath(name: string): string {
    return join(this.dataDir, name);
  }

  private readFile(name: string): CsvRow[] {
    const path = this.filePath(name);
    if (!existsSync(path)) {
      throw new Error(`Expected data file not found: ${path}`);
    }
    return loadCsv(path);
  }

  /** Brasileirao_Matches.csv — Série A 2012-2022 with rounds + team states. */
  private loadBrasileirao(): void {
    const source = MATCH_FILES.brasileirao;
    for (const row of this.readFile(source)) {
      this.matches.push({
        competition: BRASILEIRAO,
        source,
        date: parseDate(row.datetime),
        dateRaw: row.datetime ?? null,
        season: parseIntOrNull(row.season),
        round: emptyToNull(row.round),
        stage: null,
        homeTeam: cleanTeamName(row.home_team ?? ""),
        awayTeam: cleanTeamName(row.away_team ?? ""),
        homeTeamRaw: row.home_team ?? "",
        awayTeamRaw: row.away_team ?? "",
        homeGoals: parseGoals(row.home_goal),
        awayGoals: parseGoals(row.away_goal),
      });
    }
  }

  /** Brazilian_Cup_Matches.csv — Copa do Brasil. */
  private loadCup(): void {
    const source = MATCH_FILES.cup;
    for (const row of this.readFile(source)) {
      this.matches.push({
        competition: COPA_DO_BRASIL,
        source,
        date: parseDate(row.datetime),
        dateRaw: row.datetime ?? null,
        season: parseIntOrNull(row.season),
        round: emptyToNull(row.round),
        stage: emptyToNull(row.round),
        homeTeam: cleanTeamName(row.home_team ?? ""),
        awayTeam: cleanTeamName(row.away_team ?? ""),
        homeTeamRaw: row.home_team ?? "",
        awayTeamRaw: row.away_team ?? "",
        homeGoals: parseGoals(row.home_goal),
        awayGoals: parseGoals(row.away_goal),
      });
    }
  }

  /** Libertadores_Matches.csv — Copa Libertadores with tournament stage. */
  private loadLibertadores(): void {
    const source = MATCH_FILES.libertadores;
    for (const row of this.readFile(source)) {
      this.matches.push({
        competition: LIBERTADORES,
        source,
        date: parseDate(row.datetime),
        dateRaw: row.datetime ?? null,
        season: parseIntOrNull(row.season),
        round: null,
        stage: emptyToNull(row.stage),
        homeTeam: cleanTeamName(row.home_team ?? ""),
        awayTeam: cleanTeamName(row.away_team ?? ""),
        homeTeamRaw: row.home_team ?? "",
        awayTeamRaw: row.away_team ?? "",
        homeGoals: parseGoals(row.home_goal),
        awayGoals: parseGoals(row.away_goal),
      });
    }
  }

  /** BR-Football-Dataset.csv — extended stats, tournament column drives competition. */
  private loadExtended(): void {
    const source = MATCH_FILES.extended;
    for (const row of this.readFile(source)) {
      const date = parseDate(row.date);
      this.matches.push({
        competition: (row.tournament ?? "").trim() || "Unknown",
        source,
        date,
        dateRaw: row.date ?? null,
        season: date ? date.getUTCFullYear() : null,
        round: null,
        stage: null,
        homeTeam: cleanTeamName(row.home ?? ""),
        awayTeam: cleanTeamName(row.away ?? ""),
        homeTeamRaw: row.home ?? "",
        awayTeamRaw: row.away ?? "",
        homeGoals: parseGoals(row.home_goal),
        awayGoals: parseGoals(row.away_goal),
      });
    }
  }

  /** novo_campeonato_brasileiro.csv — historical Série A 2003-2019. */
  private loadHistorical(): void {
    const source = MATCH_FILES.historical;
    for (const row of this.readFile(source)) {
      this.matches.push({
        competition: BRASILEIRAO,
        source,
        date: parseDate(row.Data),
        dateRaw: row.Data ?? null,
        season: parseIntOrNull(row.Ano),
        round: emptyToNull(row.Rodada),
        stage: null,
        homeTeam: cleanTeamName(row.Equipe_mandante ?? ""),
        awayTeam: cleanTeamName(row.Equipe_visitante ?? ""),
        homeTeamRaw: row.Equipe_mandante ?? "",
        awayTeamRaw: row.Equipe_visitante ?? "",
        homeGoals: parseGoals(row.Gols_mandante),
        awayGoals: parseGoals(row.Gols_visitante),
      });
    }
  }

  /** fifa_data.csv — player database. */
  private loadPlayers(): void {
    for (const row of this.readFile(PLAYER_FILE)) {
      const id = parseIntOrNull(row.ID);
      if (id === null) continue;
      this.players.push({
        id,
        name: (row.Name ?? "").trim(),
        age: parseIntOrNull(row.Age),
        nationality: (row.Nationality ?? "").trim(),
        overall: parseIntOrNull(row.Overall),
        potential: parseIntOrNull(row.Potential),
        club: (row.Club ?? "").trim(),
        position: (row.Position ?? "").trim(),
        jerseyNumber: parseIntOrNull(row["Jersey Number"]),
        height: (row.Height ?? "").trim(),
        weight: (row.Weight ?? "").trim(),
        preferredFoot: (row["Preferred Foot"] ?? "").trim(),
        value: (row.Value ?? "").trim(),
        wage: (row.Wage ?? "").trim(),
      });
    }
  }
}

function emptyToNull(value: string | undefined): string | null {
  if (value === undefined) return null;
  const trimmed = value.trim();
  return trimmed === "" ? null : trimmed;
}

/** Resolve the default data directory (…/data/kaggle) relative to a base. */
export function defaultDataDir(baseDir: string): string {
  return join(baseDir, "data", "kaggle");
}
